mountpoint-s3/src/inode.rs

```diff
@@ -354,6 +356,66 @@ impl Superblock {
 
         Ok(LookedUp { inode, stat })
     }
+
+    /// Remove local-only empty directory, i.e., the ones created by mkdir.
+    /// It does not affect empty directories represented remotely with directory markers.  
+    pub async fn rmdir<OC: ObjectClient>(
+        &self,
+        client: &OC,
+        parent_ino: InodeNo,
+        name: &OsStr,
+    ) -> Result<(), InodeError> {
+        let LookedUp { inode, .. } = self.lookup(client, parent_ino, name).await?;
+
+        if inode.kind() == InodeKind::File {
+            return Err(InodeError::NotADirectory(inode.ino()));
+        }
+
+        let parent = self.inner.get(parent_ino)?;
+        let mut parent_state = parent.get_mut_inode_state()?;
+
+        let mut inode_state = inode.get_mut_inode_state()?;
+
+        match &inode_state.write_status {
+            WriteStatus::LocalOpen => unreachable!("A directory cannot be in Local open state"),
+            WriteStatus::Remote => {
+                return Err(InodeError::CannotRemoveRemoteDirectory(inode.ino()));
+            }
+            WriteStatus::LocalUnopened => match &mut inode_state.kind_data {
+                InodeKindData::File {} => unreachable!("Already checked that inode is a directory"),
+                InodeKindData::Directory {
+                    writing_children,
+                    deleted,
+                    ..
+                } => {
+                    if !writing_children.is_empty() {
+                        return Err(InodeError::DirectoryNotEmpty(inode.ino()));
+                    }
+                    *deleted = true;
+                }
+            },
+        }
+
+        match &mut parent_state.kind_data {
+            InodeKindData::File {} => {
+                debug_assert!(false, "inodes never change kind");
+                return Err(InodeError::NotADirectory(parent.ino()));
+            }
+            InodeKindData::Directory {
+                children,
+                writing_children,
+                ..
+            } => {
+                debug_assert!(
+                    writing_children.remove(&inode.ino()),
```

Comment:
nit: I would move the action out of the assert to make it more visible. And just store the result in a bool, e.g.: `let removed = ...`